build(docker): add web client and control center services to docker compose setup

prasad89 opened the pull request:
## Type of Change
<!-- Put an `x` in the boxes that apply -->

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Added web client and control center services to docker-compose.yml
REF : #4110 


### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [x] This PR modifies application configuration/environment variables

SanchithHegde commented on docker-compose.yml:
```diff
@@ -290,19 +323,19 @@ services:
       KAFKA_TRANSACTION_STATE_LOG_MIN_ISR: 1
       KAFKA_TRANSACTION_STATE_LOG_REPLICATION_FACTOR: 1
-      KAFKA_PROCESS_ROLES: 'broker,controller'
+      KAFKA_PROCESS_ROLES: "broker,controller"
```
Could you revert these formatting changes?

prasad89 replied:
Reverted :+1:

SanchithHegde reviewed (commented):
Did you test the changes in this PR? How does the setup process look like?

prasad89 commented:
> Did you test the changes in this PR? How does the setup process look like?

Yes, I have tested changes locally by [following](https://github.com/juspay/hyperswitch/blob/main/docs/try_local_system.md)
after doing `docker compose up -d` I'm able to see UI with *Hyperswitch Unified Checkout* text on page.

SanchithHegde commented:
> Yes, I have tested changes locally by [following](https://github.com/juspay/hyperswitch/blob/main/docs/try_local_system.md) after doing `docker compose up -d` I'm able to see UI with **Hyperswitch Unified Checkout** text on page.

If you're seeing only the text on the page, then the setup in incomplete, you should be able to see a demo store on the left, and a form to enter card details on the right as well, if you set it up correctly. Could you open the network tab in the browser developer tools and fix any problems that you notice?

prasad89 commented:
> > Yes, I have tested changes locally by [following](https://github.com/juspay/hyperswitch/blob/main/docs/try_local_system.md) after doing `docker compose up -d` I'm able to see UI with **Hyperswitch Unified Checkout** text on page.
> 
> If you're seeing only the text on the page, then the setup in incomplete, you should be able to see a demo store on the left, and a form to enter card details on the right as well, if you set it up correctly. Could you open the network tab in the browser developer tools and fix any problems that you notice?

Thank you for guiding me in the right direction. 
I've checked the network tab and fixed errors by adding the following block in the `hyperswitch-web` section:

```yaml
- HYPERSWITCH_PUBLISHABLE_KEY=PUBLISHABLE_KEY
- HYPERSWITCH_SECRET_KEY=SECRET_KEY
- HYPERSWITCH_SERVER_URL=http://hyperswitch-server:8080
- HYPERSWITCH_CLIENT_URL=http://hyperswitch-web:9050
- SELF_SERVER_URL=http://hyperswitch-web:5252
```

I've also created an API key as per the [instructions](https://github.com/juspay/hyperswitch-web/blob/main/README.md). 
However, I'm encountering an error:

```json
{
    "error": {
        "type": "invalid_request",
        "message": "API key not provided or invalid API key used",
        "code": "IR_01"
    }
}
```

I've double-checked everything from the key to matching it inside the container by inspecting the `env`, and everything seems fine to me. 
Could it be that newly created keys take some time to be fetched?

prasad89 commented:
> Did you test the changes in this PR? How does the setup process look like?

To set up, create a `.env` file and configure the following keys:

```plaintext
HYPERSWITCH_PUBLISHABLE_KEY=your_publishable_key
HYPERSWITCH_SECRET_KEY=your_secret_key
HYPERSWITCH_SERVER_URL=http://hyperswitch-server:8080
HYPERSWITCH_CLIENT_URL=http://hyperswitch-web:9050
SELF_SERVER_URL=http://hyperswitch-web:5252
```

Then execute `docker-compose up -d`.

SanchithHegde commented:
> After this changes I'm able to see the demo store on the left and a form to enter card details on the right.

Could you please include screenshots for the same?

prasad89 commented:
> > After this changes I'm able to see the demo store on the left and a form to enter card details on the right.
> 
> Could you please include screenshots for the same?

![Screenshot from 2024-03-27 19-11-42](https://github.com/juspay/hyperswitch/assets/67261499/88bc8267-2b75-4e1e-b818-81e5791dd68f)

prasad89 commented:
Looks like I need to further investigate and fix environment variables for the web.
HyperLoader is still not functioning, but I have managed to make the control center functional.
![Screenshot from 2024-03-27 23-02-25](https://github.com/juspay/hyperswitch/assets/67261499/a94d2286-e6b4-4061-a2d1-763460230d0d)

prasad89 commented:
I got the card details in right hand side now but still there some errors in network tab, looking into it.
![Screenshot from 2024-03-28 00-13-30](https://github.com/juspay/hyperswitch/assets/67261499/aa28f9fe-34e1-4ddc-a6fc-241098604dcf)

prasad89 commented:
Hello @SanchithHegde can you review the changes?

SanchithHegde commented on docker-compose.yml:
```diff
@@ -122,6 +122,45 @@ services:
+    environment:
+      - HYPERSWITCH_PUBLISHABLE_KEY=$HYPERSWITCH_PUBLISHABLE_KEY
+      - HYPERSWITCH_SECRET_KEY=$HYPERSWITCH_SECRET_KEY
+      - HYPERSWITCH_SERVER_URL=${HYPERSWITCH_SERVER_URL:-http://hyperswitch-server:8080}
+      - HYPERSWITCH_CLIENT_URL=${HYPERSWITCH_CLIENT_URL:-http://localhost:9050}
+      - SELF_SERVER_URL=${SELF_SERVER_URL:-http://localhost:5252}
```
I believe using `http://hyperswitch-server:8080` for the server URL will cause the application to be inaccessible when making payments using the web client from the browser. Can you check if that is the case?

SanchithHegde commented on docker-compose.yml:
```diff
@@ -122,6 +122,45 @@ services:
+      - "9000:9000"
+    networks:
+      - router_net
+    environment:
+      - apiBaseUrl=http://localhost:8080
+      - sdkBaseUrl=http://localhost:9050
```
Similar thing here, the control center will not be able to access either services if it is configured to use these URLs.

lsampras replied:
@SanchithHegde these url's are not used by the control center itself instead the browser will be connecting to these URL's to fetch the relavant data, so should be fine to use localhost

lsampras commented on docker-compose.yml:
```diff
@@ -122,6 +122,45 @@ services:
       logs: "promtail"
 
+### Web Client
+  hyperswitch-web:
```
also can we add 
```yaml
    labels:
      logs: "promtail"
```

here to allow promtail to scrape these logs?

prasad89 replied:
Sure

lsampras commented on docker-compose.yml:
```diff
@@ -122,6 +122,45 @@ services:
+        condition: service_started
+
+  ### Control Center
+  hyperswitch-control-center:
```
also can we add 
```yaml
    labels:
      logs: "promtail"
```

here to allow promtail to scrape these logs?

lsampras commented on docker-compose.yml:
```diff
@@ -122,6 +122,45 @@ services:
+    image: juspaydotin/hyperswitch-control-center
+    ports:
+      - "9000:9000"
+    networks:
+      - router_net
```
this might not be needed since there isn't any inter-service communication happening here...
since all communication happens on the host machine and is initiated by browser.

prasad89 replied:
![image](https://github.com/juspay/hyperswitch/assets/67261499/d5d1ec3a-6c1f-4fa5-8d1a-5e2ce3d6520e)
@lsampras have tested the flow added one fix for the same.

prasad89 commented on docker/web.Dockerfile:
```diff
@@ -0,0 +1,18 @@
+
+RUN git clone https://github.com/juspay/hyperswitch-web .
+
+RUN sed -i '/hot: true,/a \  host: "0.0.0.0",' webpack.dev.js
+RUN sed -i '/hot: true,/a \  host: "0.0.0.0",' Hyperswitch-React-Demo-App/webpack.dev.js
```
No

prasad89 commented on docker-compose.yml:
```diff
@@ -122,6 +122,45 @@ services:
+
+  ### Control Center
+  hyperswitch-control-center:
+    image: juspaydotin/hyperswitch-control-center
```
Okay, will make the changes

lsampras commented on docker-compose.yml:
```diff
@@ -122,6 +122,45 @@ services:
+    env_file:
+      - ./.env
+    environment:
+      - HYPERSWITCH_PUBLISHABLE_KEY=$HYPERSWITCH_PUBLISHABLE_KEY
+      - HYPERSWITCH_SECRET_KEY=$HYPERSWITCH_SECRET_KEY
```
@prasad89 is it necessary to pass in a publishable & secret key?
what would the behaviour be if these env variables are not set

@SanchithHegde what do you feel here about the requirement of a publishable key & secret key as part of starting up our stack?

should we move the demo app to a separate docker service or exclude it from the requirements?

prasad89 replied:
UI will not be loaded without that.

lsampras replied:
Ok,
as long as `http://localhost:9050/HyperLoader.js`(the sdk) loads without the publishable key, this should be fine then,
thanks for clearing that up

prasad89 commented on docker-compose.yml:
```diff
@@ -122,6 +122,45 @@ services:
+    build:
+      context: ./docker
+      dockerfile: web.Dockerfile
+    networks:
+      - router_net
```
Correct

lsampras commented on docker-compose.yml:
```diff
@@ -122,6 +122,45 @@ services:
+      context: ./docker
+      dockerfile: web.Dockerfile
+    env_file:
+      - ./.env
```
@prasad89 can you add this env file if it is needed, I'd prefer all variables in the below `environment` block tho?

prasad89 replied:
We've set up a configuration block with default variable pointers. 
```yaml
 - HYPERSWITCH_SERVER_URL=${HYPERSWITCH_SERVER_URL:-http://hyperswitch-server:8080}
 - HYPERSWITCH_CLIENT_URL=${HYPERSWITCH_CLIENT_URL:-http://localhost:9050}
 - SELF_SERVER_URL=${SELF_SERVER_URL:-http://localhost:5252}
```
For confidentiality, `HYPERSWITCH_PUBLISHABLE_KEY` and `HYPERSWITCH_SECRET_KEY` are considered personal secrets and should be stored in a .env file. 

Therefore, the complete .env file should appear as follows:

```yaml
HYPERSWITCH_PUBLISHABLE_KEY=HYPERSWITCH_PUBLISHABLE_KEY
HYPERSWITCH_SECRET_KEY=HYPERSWITCH_SECRET_KEY
HYPERSWITCH_SERVER_URL=http://hyperswitch-server:8080
HYPERSWITCH_CLIENT_URL=http://localhost:9050
SELF_SERVER_URL=http://localhost:5252
```

lsampras replied:
but then you are specifying these keys here as well,
```
    environment:
      - HYPERSWITCH_PUBLISHABLE_KEY=$HYPERSWITCH_PUBLISHABLE_KEY
      - HYPERSWITCH_SECRET_KEY=$HYPERSWITCH_SECRET_KEY
prasad89 marked this conversation as resolved.
      - HYPERSWITCH_SERVER_URL=${HYPERSWITCH_SERVER_URL:-http://hyperswitch-server:8080}
      - HYPERSWITCH_CLIENT_URL=${HYPERSWITCH_CLIENT_URL:-http://localhost:9050}
      - SELF_SERVER_URL=${SELF_SERVER_URL:-http://localhost:5252}
```

in that case we can remove the .env file

prasad89 replied:
In our configuration, we're defining the `environment` block and specifying that `HYPERSWITCH_PUBLISHABLE_KEY` and `HYPERSWITCH_SECRET_KEY` are required for running Docker Compose, indicated by the presence of the `'$` symbol. For other variables, Docker Compose will use values from the `.env` file if provided, otherwise, it will proceed with default value from `environment` block

We are using `interpolation` method here.
REF : https://docs.docker.com/compose/environment-variables/env-file/#interpolation

lsampras replied:
```
environment:
  - HYPERSWITCH_PUBLISHABLE_KEY=$HYPERSWITCH_PUBLISHABLE_KEY
  - HYPERSWITCH_SECRET_KEY=$HYPERSWITCH_SECRET_KEY
```

oh I was confused by these part then? where would this be used?

prasad89 replied:
Yes, those should be present in the `.env` file in order to run docker compose.

prasad89 replied:
@lsampras hope I explained you well.
Let me know if there are anymore changes.

lsampras replied:
yeah, 
can you make the env file optional as mentioned [here](https://docs.docker.com/compose/environment-variables/set-environment-variables/#additional-information-2).
Also it seemed that control-center was facing some problems when loading the sdk for test payment..
<img width="1422" alt="image" src="https://github.com/juspay/hyperswitch/assets/21202349/3b79cf8a-fb29-4ee0-aef5-857db12afe44">

Can you test that flow once on your end and lmk if you are able to make payments through that?
you can configure a test connector for the same.

prasad89 replied:
@lsampras have tested the flow added one fix for the same. 
![Screenshot from 2024-04-26 20-04-43](https://github.com/juspay/hyperswitch/assets/67261499/943b54e2-9875-4e39-9039-74b1ea2ade37)

lsampras replied:
@prasad89 Can you make sure the SDK loads in control-center
sharing a reference screenshot from our sandbox environment (app.hyperswitch.io) 


<img width="1728" alt="image" src="https://github.com/juspay/hyperswitch/assets/21202349/e6ac6b85-8318-48c6-a740-c8741782e9db">

lsampras commented on docker-compose.yml:
```diff
@@ -122,6 +122,43 @@ services:
+      - HYPERSWITCH_PUBLISHABLE_KEY=$HYPERSWITCH_PUBLISHABLE_KEY
+      - HYPERSWITCH_SECRET_KEY=$HYPERSWITCH_SECRET_KEY
+      - HYPERSWITCH_SERVER_URL=${HYPERSWITCH_SERVER_URL:-http://hyperswitch-server:8080}
+      - HYPERSWITCH_CLIENT_URL=${HYPERSWITCH_CLIENT_URL:-http://localhost:9050}
```
@prasad89 this is missing some env variables (that are not present in the SDK documentation),
can you refer to [this issue](https://github.com/juspay/hyperswitch-web/issues/331) and add them here?

lsampras reviewed (commented):
Some minor changes,
also can you share a screenshot of the SDK loading locally (inside the control-center) as shown in the reference screenshot from sandbox

prasad89 commented on docker/web.Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+WORKDIR /hyperswitch-web
+
+RUN git clone https://github.com/juspay/hyperswitch-web .
```
Addressed

prasad89 commented on docker-compose.yml:
```diff
@@ -122,6 +122,43 @@ services:
+      - HYPERSWITCH_SERVER_URL=${HYPERSWITCH_SERVER_URL:-http://hyperswitch-server:8080}
+      - HYPERSWITCH_CLIENT_URL=${HYPERSWITCH_CLIENT_URL:-http://localhost:9050}
+      - SELF_SERVER_URL=${SELF_SERVER_URL:-http://localhost:5252}
+    depends_on:
+      hyperswitch-server:
+        condition: service_started
```
That's the nice idea.

prasad89 commented:
@lsampras, I've addressed your review. Could you please review it? 
Additionally, I'm encountering a `401` unauthorized status when attempting to load the SDK.
I've created both `HYPERSWITCH_PUBLISHABLE_KEY` and `HYPERSWITCH_SECRET_KEY`, and exported them accordingly. 
I can confirm that `HYPERSWITCH_PUBLISHABLE_KEY` is visible at `localhost:5252/config`. 
Could you provide any guidance on how to proceed? 
Your assistance would be greatly appreciated.

lsampras commented on docker-compose.yml:
```diff
@@ -137,9 +137,10 @@ services:
-    depends_on:
-      hyperswitch-server:
-        condition: service_started
+      - SDK_ENV=${SDK_ENV:-local}
+      - ENV_SDK_URL=${ENV_SDK_URL:-http://localhost:9050/HyperLoader.js}
+      - ENV_BACKEND_URL=${ENV_BACKEND_URL:-http://localhost:8080}
+      - ENV_LOGGING_URL=${ENV_LOGGING_URL:-http://localhost:3100}
```
I believe you picked this up from the issue that was mentioned in the fix, but that change hasn't been merged to `main` branch yet,

either use the older SDK variable keys or pull that branch instead of main in your dockerfile

lsampras commented on docker/web.Dockerfile:
```diff
@@ -4,12 +4,12 @@ RUN npm install concurrently -g
 WORKDIR /hyperswitch-web
 
-RUN git clone https://github.com/juspay/hyperswitch-web .
+RUN git clone https://github.com/juspay/hyperswitch-web . --depth 1
```
The ordering here seems incorrect

lsampras commented on docker-compose.yml:
```diff
@@ -137,9 +137,10 @@ services:
-      hyperswitch-server:
-        condition: service_started
+      - SDK_ENV=${SDK_ENV:-local}
+      - ENV_SDK_URL=${ENV_SDK_URL:-http://localhost:9050/HyperLoader.js}
```
This won't have the `HyperLoader.js` at the end this is supposed to be the base url only

lsampras commented on docker-compose.yml:
```diff
@@ -137,10 +137,6 @@ services:
-      - SDK_ENV=${SDK_ENV:-local}
-      - ENV_SDK_URL=${ENV_SDK_URL:-http://localhost:9050/HyperLoader.js}
-      - ENV_BACKEND_URL=${ENV_BACKEND_URL:-http://localhost:8080}
-      - ENV_LOGGING_URL=${ENV_LOGGING_URL:-http://localhost:3100}
```
can you add these variables back?

prasad89 replied:
Okay, will add it.

I'm getting this errors from morning on `hyperswitch-router:standalone`

```Unable to deserialize application configuration: connector_customer.payout_connector_list: Some errors occurred:
Unable to deserialize `stripe` as `api_models::enums::PayoutConnectors`: Matching variant not found
thread 'main' panicked at crates/router/src/bin/router.rs:14:10:
Unable to construct application configuration: ConfigurationError(Some errors occurred:
Unable to deserialize `stripe` as `api_models::enums::PayoutConnectors`: Matching variant not found)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace```

@lsampras do you have any idea regarding this?

SanchithHegde replied:
Oh, this seems to be caused by recent changes to the `docker_compose.toml` file.

I would have suggested checking out to the `v1.107.0` tag before running the Docker Compose setup, but since you need to open a PR against the `main` branch, that isn't an option. You can remove `stripe` from the `payout_connector_list` line in the [`docker_compose.toml` file](https://github.com/juspay/hyperswitch/blob/b0133f33693575f2145d295eff78dd07b61efcda/config/docker_compose.toml#L398) to be able to run the Docker Compose setup, and you need not push the `docker_compose.toml` change as part of this PR.

prasad89 replied:
Thanks for the help @SanchithHegde 
Also when I'm trying to preview payment from control center UI is authenticating request with `beta.hyperswitch.io` is that expected?
Should I use key created there.

prasad89 commented:
I'm getting this errors on `hyperswitch-router:standalone`

```Unable to deserialize application configuration: connector_customer.payout_connector_list: Some errors occurred:
Unable to deserialize `stripe` as `api_models::enums::PayoutConnectors`: Matching variant not found
thread 'main' panicked at crates/router/src/bin/router.rs:14:10:
Unable to construct application configuration: ConfigurationError(Some errors occurred:
Unable to deserialize `stripe` as `api_models::enums::PayoutConnectors`: Matching variant not found)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

@SanchithHegde do you have any idea regarding this?

lsampras reviewed (approved):
Can get this merged for now, seems good for the most part

prasad89 commented:
Thanks @SanchithHegde and @lsampras for your time and approving this PR.
This will keep me motivating to contribute further.
